### ISSUE TITLE
The pattern matcher do not guess groundable clause present in conditional clause

### ISSUE THREAD

**Issue by ngeiswei** [opencog/atomspace#1420]
Currently the pattern matcher returns a user error on
```
(Bind
  (Evaluation
    (GroundedPredicate "scm: true-enough")
    (Inheritance (Variable "$X") (Variable "$Y"))
  (Concept "OK"))
```
given the grounded predicate
```
(define (true-enough-bool a)
  (let ((s (cog-stv-strength a)) (c (cog-stv-confidence a)))
    (and (> s 0.5) (> c 0.5))))
```
returns the user error
```
ERROR: In procedure cog-new-link:
ERROR: Throw to key `C++-EXCEPTION' with args `("cog-new-link" "Expecting an outgoing set size of at most two, got 1 (/home/nilg/OpenCog/atomspace/opencog/atoms/pattern/BindLink.cc:85)\nFunction args:\n((EvaluationLink\n   (GroundedPredicateNode \"scm: true-enough\")\n   (InheritanceLink\n      (VariableNode \"$X\")\n      (VariableNode \"$Y\")\n   )\n   (ConceptNode \"OK\")\n)\n)")'.
```
while it could probably unambiguously guess that `(Inheritance (Variable "$X") (Variable "$Y"))` is a groundable close.

**Comment by linas** [comment]
Yes, this is discussed in greater detail in a post on the mailing list.

**Comment by linas** [comment]
I can try to fix, but not soon.

**Comment by linas** [comment]
So, first of all, there is a misplaced paren in the above.  I think the intent was this:
```
(Bind
    (Evaluation
        (GroundedPredicate "scm: true-enough")
        (Inheritance (Variable "$X") (Variable "$Y")))
    (Concept "OK"))
```
this throws a different error:
```
ERROR: In procedure cog-new-link:
Throw to key `C++-EXCEPTION' with args `("cog-new-link" "Variable not groundable: (VariableNode \"$Y\") ; [5319542295444993445][1]\n\n (/home/linas/src/novamente/src/atomspace-git/opencog/atoms/pattern/PatternLink.cc:849)\nFunction args:\n((EvaluationLink\n   (GroundedPredicateNode \"scm: true-enough\")\n   (InheritanceLink\n      (VariableNode \"$X\")\n      (VariableNode \"$Y\")\n   )\n)\n (ConceptNode \"OK\")\n)")'.
```

which seems ... wrong ...

**Comment by linas** [comment]
Also more typos: should have been
```
(define (true-enough a)
  (let ((s (cog-mean a)) (c (cog-confidence a)))
    (if (and (> s 0.5) (> c 0.5)) (stv 1 1) (stv 0 1) ))
```

Maybe we should change what GPN's accept ....

**Comment by linas** [comment]
@ngeiswei Can you clarify the intent of the above?  There are two (or three) possibilities:

1) implicitly search for the presence of `(Inheritance (Variable "$X") (Variable "$Y"))`
2) do NOT search for `(Inheritance (Variable "$X") (Variable "$Y"))` but only search for X and Y, and then create `(Inheritance (Variable "$X") (Variable "$Y"))` on the fly (in a scratch atomspace) passing it to the GPN
3) throw error, complaining about ill-formed expression.

**Comment by linas** [comment]
Closing. Fixed in #2662

**Comment by ngeiswei** [comment]
> That leaves option 3 as the only viable option.

I agree!
